### ISSUE TITLE
question about forward pass

### ISSUE THREAD

**Issue by srn284** [jiaqima/SODEN#5]
Hi SODEN team,
Amazing work for survival modelling and something that shows notably better performance as compared to benchmark models, @jiaqima.

I am trying to extract lambda, Lambda and the hazard outputs in training...

`class NonCoxFuncModel(nn.Module):

    def __init__(self, config):
        Arguments:
          config: An OrderedDict of lists. The keys of the dict indicate
            the names of different parts of the model. Each value of the dict
            is a list indicating the configs of layers in the corresponding
            part. Each element of the list is a list [layer_type, arguments],
            where layer_type is a string and arguments is a dict.
          feature_size: Feature size.
          use_embed: Whether to use embedding layer after input.
        """
        super(NonCoxFuncModel, self).__init__()
        self.config = config
        self.feature_size = config.hidden_size
        self.func_type = config.func_type

        self.odefunc = ContextRecMLPODEFunc(config)
        self.model_config = config
        self.set_last_eval(False)

    def set_last_eval(self, last_eval=True):
        self.last_eval = last_eval

    def forward(self, orig_t, orig_init_cond,orig_features, fullEval,labelfloat ):
        device = next(self.parameters()).device
        t = orig_t
        init_cond = orig_init_cond
        features = orig_features
        init_cond = torch.cat([orig_init_cond.view(-1, 1), orig_t.view(-1, 1), orig_features],
                              dim=1)
        t = torch.tensor([0., 1.]).to(device)

        outputs = {}
        self.odefunc.set_batch_time_mode(False)
        intermdout = odeint(
            self.odefunc, init_cond, t, rtol=1e-4, atol=1e-8)
        
        outputs["Lambda"] =intermdout[1:].squeeze()  # size: [length of t] x [batch size] x [dim of y0]
        self.odefunc.set_batch_time_mode(True)
        
        outputs["lambda"] = self.odefunc(t[1:], outputs["Lambda"]).squeeze()
        outputs["Lambda"] = outputs["Lambda"][:, 0]
        outputs["lambda"] = outputs["lambda"][:, 0] / orig_t

        self.odefunc.set_batch_time_mode(False)
        ones = torch.ones_like(orig_t)
        
        t = self.model_config.time_nums * ones
        init_cond = orig_init_cond
        features = orig_features
        init_cond = torch.cat(
            [init_cond.view(-1, 1), t.view(-1, 1), features],
            dim=1)
        t_min = 1
        t_max = self.model_config.time_nums
        maxsteps = t_max
        t = torch.linspace(
            t_min, t_max, maxsteps, dtype=init_cond.dtype,
            device=device)
    
        t = torch.cat([torch.zeros([1]).to(device), t], dim=0)
        t = t / t_max
        
        outputs["hazard_seq"] =  (1 - torch.exp (-odeint(self.odefunc, init_cond, t, rtol=1e-4,
                    atol=1e-8))[1:, :, 0]   ).transpose(1,0)

        return outputs

    `

When I only do training with the code up to line:

`outputs["lambda"] = outputs["lambda"][:, 0] / orig_t` 
 the training and evaluation look good with range of predicted probability from 0.03 to 0.999 for probabilities predicted at certain time point t for example. 

However, when I add the lines of code for the extraction of the hazard_seq - i.e. the lines of code after the above line
which ultimately yields 1- exp(-odeint output) - there is some problems with the predicted outcomes... For examples the minimum survival probability at some time point t is now 0.11 as opposed to 0.03 as said above.

I am not even doing anything with this hazard_seq in the loss function but just doing the following loss function as in the code below:

`

class SurvODEloss(torch.nn.Module):

    def __init__(self, reduction):
        super(SurvODEloss, self).__init__()
        self.reduction = reduction

    
    def forward(self, outputs, labels):
        return self.survodeloss( outputs, labels)

   
    def survodeloss(self, outputs, labels):
        def _reduction(loss, reduction):
            if reduction == 'none':
                return loss
            elif reduction == 'mean':
                return loss.mean()
            elif reduction == 'sum':
                return loss.sum()
            raise ValueError(f"`reduction` = {reduction} is not valid. Use 'none', 'mean' or 'sum'.")

        batch_loss = -labels * torch.log(
            outputs["lambda"].clamp(min=1e-8)) + outputs["Lambda"]
        return _reduction(batch_loss, self.reduction)

    `

Is there something wrong perhaps with calling odeint for the lambda outputs AND the hazard_Seq outputs? or something else I am doing wrong... Ultimately, i need to extract the cdf at a certain time point (which i need to do some post-processing which is fine), and use this cdf as input for a regularisation scheme. Hence, asking this question. Thank you very much for the time and efforts!

PS. dropped an email your way too regarding this. please feel free to reply there too if more convenient. thank you @jiaqima!
